Turbo refactoring

paymicro opened the pull request:
# TODO
- [ ] “Do not automatically scroll down while receiving the streaming, as it forces us to scroll up to read the beginning of the message.”
- [x] “Reintroduce the ability to edit the chat name on the history screen.”
- [ ] “When opening an existing chat, ideally position the scroll at the beginning of the last response, so it's not necessary to scroll up to read the last message. This is already the case in the current version.”
- [x] When clicking the “Copy as image” button on a mermaid diagram, provide feedback similar to the other commands that already do this.
- [ ] “Highlight special items: ‘/’ for commands, ‘@’ for items, and API and SQL tags within the chat, just like in the current version.”
- [ ] Referencing parts of the project using “@” stopped working. With “@” in the request field, the window to reference parts of the project is not being displayed.
 ***
- [x] Testing / fixing bugs in history
- [x] Testing in light theme
- [x] Testing in VS 2019... 😒 
- [x] Use JS scripts from local resource
- [x] Add back the buttons `Apply` and `Copy` on code blocks
- [x] Add mermaid renderer and `Copy` button
- [x] Do not automatically scroll down while receiving the streaming (or scrolling only when view already scrolled)
- [x] Reintroduce the ability to edit the chat name on the history screen.
- [x] Allow closing the side window by clicking outside of it, instead of always having to click the "X".
- [x] When clicking the copy button, provide feedback for the click, for example, display a check icon for a few seconds, as happens in the current version.
- [x] When opening an existing chat, ideally position the scroll at the beginning of the last response, so it's not necessary to scroll up to read the last message. This is already the case in the current version.
- [x] The Apply button still doesn't work.
- [x] Highlight special items: "/" for commands, "@" for items, and API and SQL tags within the chat, just like in the current version.
- [x] There's a bug when adding API and database contexts to the chat - it ends up adding them to all chats.
- [x] Chats are still not receiving a title from the AI.
- [x] Exception when clicking delete some history them and received. Because the chat wasn't found in the database for deletion.
- [x] Functions are not working. It seems the function definitions are not being sent.
________
**Preview**

VS 2022
<img width="721" height="675" alt="image" src="https://github.com/user-attachments/assets/ecefecb7-c72a-4c07-a17d-13a9069d891f" />

VS 2026
<img width="698" height="770" alt="image" src="https://github.com/user-attachments/assets/65da9370-d62a-44ed-ada3-d30f7e88918d" />
<img width="727" height="731" alt="image" src="https://github.com/user-attachments/assets/ff4053fc-b2e6-423e-9d4c-ae25e79ddea8" />
<img width="693" height="620" alt="image" src="https://github.com/user-attachments/assets/7f25f04d-8eeb-4cbc-ab6a-2b47fad2d431" />

paymicro committed:
turbo UI refactoring

paymicro committed:
turbo refactoring part 2

paymicro committed:
fix history slidebar

paymicro committed:
fix history paging

paymicro committed:
fix UI

paymicro committed:
delete chat in history

paymicro committed:
stream option + colors

paymicro committed:
added settings slidebar

paymicro committed:
Merge remote-tracking branch 'upstream/master' into turbo-refactoring

paymicro committed:
catch completion command exception

paymicro committed:
update gpt bubble bg color

paymicro committed:
cancel stream support

jeffdapaz commented on VisualChatGPTStudioShared/ToolWindows/Turbo/TerminalWindowTurboControl.xaml.cs:
```diff
@@ -33,373 +90,1417 @@ public partial class TerminalWindowTurboControl : UserControl
+
+                cancellationTokenSource = new();
+
+                if (options.CompletionStream)
```
It seems that when using streaming, function calls are being ignored. However, there is a way to capture function calls even with streaming.

After receiving all the chunks, you only need to follow the steps below to retrieve the function calls:

```C
if (ApiChat.MostRecentApiResult?.Choices != null && ApiChat.MostRecentApiResult.Choices.Any())
    {
        var choice = ApiChat.MostRecentApiResult.Choices.First();
        if (choice?.Functions != null && choice.Functions.Any())
        {
            functions = choice.Functions.ToList();
        }
    }
```

jeffdapaz commented on VisualChatGPTStudioShared/ToolWindows/Turbo/TerminalWindowTurboControl.xaml.cs:
In the last release, I added a feature that allows users to preview mermaid diagrams by clicking a new button:

<img width="1377" height="634" alt="image" src="https://github.com/user-attachments/assets/657ebf33-48d3-49e2-9bbd-6de1fc279b67" />

I was thinking of changing this. Instead of opening the code in a new window to render the diagram using the VS render feature, maybe render the diagram directly within the chat itself, just below the Mermaid code.

Since you're already involved in all these refactors, perhaps you could take this on as well 🙂

jeffdapaz replied:
Very good! Thank you. 

But one detail, it would be great if the user could visualize and/or copy the Mermaid code, as they might want to use or share it in some way.

paymicro commented on VisualChatGPTStudioShared/Options/OptionPageGrid.cs:
```diff
@@ -244,6 +244,12 @@ public class OptionPageGridGeneral : DialogPage
+        [DisplayName("Stream")]
+        [Description("Change the Stream option if you wish to stream response")]
+        [DefaultValue(false)]
+        public bool CompletionStream { get; set; } = false;
```
I'm not 100% sure about this, so it's better to leave the setting

paymicro committed:
restore GetResponseFromChatbotAsync method

jeffdapaz commented on VisualChatGPTStudioShared/ToolWindows/Turbo/TerminalWindowTurboControl.xaml.cs:
```diff
@@ -33,373 +90,1417 @@ public partial class TerminalWindowTurboControl : UserControl
+                _webView = null;
+            }
+
+            _webView = new WebView2CompositionControl();
```
The problem is the WebView2CompositionControl doens't work on VS2019.

I now this can be ugly, but maybe only for VS2019, if history, config, etc, be opened, set the webView visibility to hidden or collapsed. Maybe this workaround resolve this issue for the VS2019.

paymicro committed:
copy, apply, functions,

paymicro committed:
mermaid

paymicro committed:
copy webassets to appdata

paymicro committed:
vs2019 webassets

paymicro commented:
hi @jeffdapaz
Thanks for tests.

> Chats are still not receiving a title from the AI.

Strange. I'll check it.
_____
> What is the purpose of the "Delete Chat" button? Does it delete the chat from the repository, or does it just clear the current chat?

`Delete Chat`
- delete current chat from database 
- clear UI

I assume the behavior is that a new chat is created on the first message.
It doesn't work quite like that now.
_____

> I didn’t investigate further, but functions are not working. It seems the function definitions are not being sent.

I've never used this feature. Can you provide instructions? Is this similar to MCP client?

jeffdapaz commented:
> Strange. I'll check it.

Yeah, for me, all new chats kept the "New Chat" name, even after sending requests.

> I assume the behavior is that a new chat is created on the first message.
It doesn't work quite like that now.

Alright, I think it's because you're not finished yet, and that's why the chat context isn't clearing for new chats.

> I've never used this feature. Can you provide instructions? Is this similar to MCP client?

You can read [here](https://platform.openai.com/docs/guides/function-calling).

When you request the API to execute a function, we first need to send the functions definitions. I think the issue lies here, the definitions maybe are not being sent.

Maybe you can check the current version of the code to figure out what’s going wrong and how works now.

paymicro committed:
applyCode, highlightSpecialTags, rename history, order items in history, scrollIfNeeded, naming chats

paymicro committed:
minor fixing

paymicro committed:
scroll. tool_calls, copy popup

paymicro committed:
remove test

paymicro committed:
UI fixing

paymicro committed:
Update api key

paymicro committed:
Merge remote-tracking branch 'upstream/master' into turbo-refactoring

paymicro committed:
apply and copy selected code

jeffdapaz commented:
Hi @paymicro ,

Just letting you know that I'm reviewing your incoming commits, and I'm waiting for you to finish the TODOs to test again.

paymicro commented:
Hi @jeffdapaz 
I'm ready to review.

paymicro committed:
minor JS fixes

jeffdapaz commented:
Hi @paymicro ,

All right, I'll review and try to figure out what's going on with VS2019. I'll provide feedback later.

jeffdapaz commented:
@paymicro ,

I had the opportunity to run some tests. Here’s my feedback:

1 - What do you intend to add in Settings? Would these be different configurations from those already possible via options?

2 - In the Chat History window, add the ability to navigate between chats (list items) using the UP and DOWN keys. Additionally, allow to select with the Enter key. This way, it will be possible to navigate using only the keyboard.

3 - In the chat (selected list item), allow to choose the "Rename" and "Delete" commands with the TAB key, and confirm and execute the button with Enter.

4 - I think it would be better to be able to select the chat (list item) with just one mouse click instead of a double click.

5 - I noticed that, for example, if we have a chat open in a "dock" that has other Visual Studio windows, and we view another window, when we return to the chat window it doesn’t display the chat that was previously open, but another one. Ideally, it should return to the chat that was open.

6 - When a new message is added to the chat (whether mine or the AI’s), it would be ideal to automatically scroll to the start of that message. Sometimes the new message is hide outside the chat.

7 - I preferred it when our messages had a green background, it made them stand out more and differentiated them from the AI’s messages 🙂

8 - By the way, a suggestion for improvement regarding mermaid diagrams would be to add a new button that allows downloading the diagram as an image. The CDN itself already has a function for this. Here’s an example:

```java
function downloadPNG() {
      const code = document.getElementById("mermaidInput").value;
      mermaid.mermaidAPI.render("graphDownload", code, (svgCode) => {
        // Converter SVG em Canvas
        const image = new Image();
        const svg = new Blob([svgCode], {type: 'image/svg+xml;charset=utf-8'});
        const url = URL.createObjectURL(svg);
        image.onload = function() {
          const canvas = document.createElement('canvas');
          canvas.width = image.width;
          canvas.height = image.height;
          const ctx = canvas.getContext('2d');
          ctx.drawImage(image, 0, 0);
          URL.revokeObjectURL(url);
          canvas.toBlob(function(blob) {
            const link = document.createElement('a');
            link.download = 'diagrama.png';
            link.href = URL.createObjectURL(blob);
            link.click();
          }, "image/png");
        };
        image.src = url;
      });
    }
```

9 - The shortcut CTRL+ALT+Enter for the "Send Code" command is not working.

10 - In the current version of the chat, when the "Send Code" button is pressed, it checks if the AI’s response contains a code block, and if it does, it applies this block to the active editor, and what is not a code block is printed in the chat window. If the "Send Code" command is pressed while holding the SHIFT key, it will not write the code block from the response to the editor, but rather in the chat. However, from what I tested, now it always writes the code block in the chat, regardless of whether I pressed the SHIFT key or not.

11 - The highlight for special items ("/" for commands, "@" for items, API and SQL tags within the chat and images) is still not working properly.

12 - I ran tests with the "Function" feature of SQL Server and agent APIs, and I see that the function is being returned and executed internally in VS. However, when we return the result to the AI, I get this error:

HTTP status code: BadRequest. Content: {
  "error": {
    "message": "Invalid parameter: messages with role 'tool' must be a response to a preceeding message with 'tool_calls'.",
    "type": "invalid_request_error",
    "param": "messages.[4].role",
    "code": null
  }
}

13 - I did some tests with the Computer Use feature. The first time it worked, but on the second attempt I got this error:

BadRequest: {
  "error": {
    "message": "No tool output found for computer call call_1jz1ksZSCK3vk89D37bgZODb.",
    "type": "invalid_request_error",
    "param": "input",
    "code": null
  }
}

It seems there’s something wrong with the "Function" and "Computer Use" requests. I haven’t investigated deeply, but I can do that when I have more time.

14 - For the "light" and "blue" themes in VS, the code and mermaid blocks still appear with the dark theme.

15 - When using the VS "blue" theme, the buttons are almost invisible.

16 - I tried testing with VS2019 but had the same error as you. When I have more time, I’ll investigate further and try to understand what’s going on.

17 - The issue with the context of APIs and SQL Server connections being shared across all chats still persists. When added, it should be exclusive to that specific chat. For example, open a chat and add a reference to an API and/or SQL Server connection. Then open another chat and try to add the same API or connection again. You’ll see a message saying that the context has already been added.

jeffdapaz commented:
> Some fast settings or specific settings for this mode:

Alright, but while this feature is still incomplete, hide the button to avoid displaying it unnecessarily.

paymicro committed:
history navigation

paymicro commented:
> For example, instead of downloading the PNG, the button copies it to the clipboard? Or save somewhere?

I'm thinking of saving it to the clipboard. Via sending to the extension, because there were problems in the browser. 
And this behavior is another candidate for placement on the Settings slider.

Now you can insert PNG from the clipboard into any editor without any problems.

jeffdapaz commented:
> > For example, instead of downloading the PNG, the button copies it to the clipboard? Or save somewhere?
> 
> I'm thinking of saving it to the clipboard. Via sending to the extension, because there were problems in the browser. And this behavior is another candidate for placement on the Settings slider.
> 
> Now you can insert PNG from the clipboard into any editor without any problems.

Alright, sounds good. One button to copy the Mermaid code and another for the PNG.

paymicro committed:
mermaid - copy as PNG

paymicro committed:
monor improvements

paymicro committed:
UI theme, VS2019

paymicro committed:
clear attachments when load chat

paymicro committed:
copy button animation

paymicro committed:
fixing bug

paymicro commented:
It looks like the work with SQL needs to be seriously refactored. I think that in any chat there was an opportunity to turn on and off SQL functions as simply as adding and deleting an image. 
To make everything more obvious. I suggest adding a page with functions. You can already turn tools on and off on it.

paymicro commented:
Something like... 🔽
<img width="978" height="971" alt="image" src="https://github.com/user-attachments/assets/dc057e8a-5254-4df2-a45f-74dea68d1cfc" />


I think that using the buttons SQL / API we can enable and disable tools (functions) and immediately send a task in the request.
This will be a mode switch.
`Normal` - no tools (+ filesystem to read files in current solution)
`API mode` -  added API tools
`SQL mode` - added SQL tools

And show function calls with result (collapsed by default)

When loading a chat there will always be normal mode without enabled functions. You can connect tools in any chat.

paymicro commented:
> 1. Would this approach still guarantee that each chat maintains its own isolated functions context, without interfering with other chats?

I assume that the set of tools can be dynamic. 
To use SQL tools, the user must explicitly enable them via a checkbox and select a connection
If the checkbox is cleared, SQL tools are excluded from the LLM context
**Transparency** - the user clearly understands which tools are active
**Security** - accidental execution of SQL queries is excluded
**Efficiency** - we don’t load LLM with unnecessary tools, we save tokens
**Control** - the user decides when and what tools to use

Other chats will not be able to influence the current one. By default, all tools will be disabled. We can store in the database which tools are enabled for this chat in the last request.

✔️ Each chat has its own tool state
✔️ On/off status is saved for each chat
✔️ Tools are disabled by default

***

> 2. With the new window for enabling/disabling functions, how is the user supposed to select which APIs and/or DB connections they want to add to the chat context?

There are buttons below the request textbox. We can make them behave like a checkbox. When you click on SQL button, it will remain in the clicked state and there will be a choice of connection to which queries will be sent.

✔️ Checkbox buttons under the input field - intuitive
✔️ The clicked state shows active tools

***

> 3. Are you also proposing any changes to how the function results will be displayed in the chat, or would that remain the same?

The results of the functions will be displayed by the model + the tool calls themselves and the results will be shown as in the screenshot above

✔️ The results are displayed by the model
✔️ Tool calls and results are visible in chat

***

> 4. I’m still a bit confused about the possibility of disabling functions once they’ve already been enabled and sent to the chat context. How do you intend to handle that scenario?

The `tools` list sended in **EVERY** user request. We can just add/remove any tool in next request. The previous tools do not affect those sent in the new request in any way. That is, each new request must contain a list of tools that the model can use.
```
curl https://api.openai.com/v1/chat/completions
  -H "Content-Type: application/json"
  -H "Authorization: Bearer $OPENAI_API_KEY"
  -d '{
    "model": "gpt-3.5-turbo",
    "messages": [
      {
        "role": "user",
        "content": "content"
      }
    ],
    "tools": [
      {
        "type": "function",
        "function": {
          "name": "get_current_weather",
          "description": "Get the current weather at a specified location",
          "parameters": {
            "type": "object",
            "properties": {
              "location": {
                "type": "string",
                "description": "City and state, such as San Francisco, California."
              },
              "unit": {
                "type": "string",
                "enum": ["celsius", "fahrenheit"]
              }
            },
            "required": ["location"]
          }
        }
      }
    ],
    "tool_choice": "auto"
}'
```
Therefore, we can open a new chat in which the tools were used and continue the chat without using them. Or we can add any in the process.

jeffdapaz commented:
>There are buttons below the request textbox. We can make them behave like a checkbox. When you click on SQL button, it will remain in the clicked state and there will be a choice of connection to which queries will be sent.
>
>✔️ Checkbox buttons under the input field - intuitive
>✔️ The clicked state shows active tools

Ok, but will be possible to add any DB connections and/or any APIs contexts to the same chat, as is possible today, right? I'm meaning, will be possible to add as many itens to chat's context the user want?

>The results of the functions will be displayed by the model + the tool calls themselves and the results will be shown as in the >screenshot above
>
>✔️ The results are displayed by the model
>✔️ Tool calls and results are visible in chat

This isn't very practical; in fact, seeing what we send and receive isn't very useful.

For example, today the results are displayed as follows:

If request any SQL action:

<img width="440" height="560" alt="image" src="https://github.com/user-attachments/assets/f1f254e2-d7d9-4c93-b1f6-33142e5e031c" />

In other words, the AI ​​requests the execution of the function, we return the result of the function to the AI, and based on that, it returns another response.

However, if the AI ​​executes a query that returns data, we don't return the result to the AI, only the number of records returned, and we display the query result in a DataGrid:

<img width="440" height="535" alt="image" src="https://github.com/user-attachments/assets/93cdb971-c0ed-4ac1-bced-89201a26bea4" />

When I said earlier that one of the problems was that the table was being displayed in all chats, this was the table I was referring to. You could switch chats, but the table would still be there.

For APIs, we have two situations. In the extension options, for each configured API, the user can choose whether or not to return the data to the AI:

<img width="1339" height="888" alt="image" src="https://github.com/user-attachments/assets/44348e61-7969-4d14-93da-f670415207ae" />

If he agrees to send the answer to the AI, it will condense the response:

<img width="546" height="744" alt="image" src="https://github.com/user-attachments/assets/f9bd657b-07c6-4a64-8ec0-3f7d547b2290" />

If not, AI only receives the status code from the APIs, without the actual responses, and you show the response directly on the chat:

<img width="510" height="572" alt="image" src="https://github.com/user-attachments/assets/83d0bfdf-6f04-462a-87e4-201728ed4e94" />

In short, I prefer this approach because the AI ​​condenses the responses, giving the user a clearer view of the results and enabling them to make requests to the AI ​​based on the function's answers.

I accept the rest of your proposal, but the way you proposed displaying the interactions with the functions, I believe we would be taking a step backward.

paymicro commented:
> Ok, but will be possible to add any DB connections and/or any APIs contexts to the same chat, as is possible today, right? I'm meaning, will be possible to add as many itens to chat's context the user want?

Sure.

> This isn't very practical; in fact, seeing what we send and receive isn't very useful.

I can't agree. But I think it will be possible to configure the invisibility of tool calls. Personally, I cannot trust the model with the right to execute scripts in SQL, even to the point of deleting tables without confirmation.
`DataGrid` can be showed separately by formatting the message in the current chat as a markdown table. The design needs a little work.

***

I would prefer this dialog to look like it does now in LM Studio. In one request, call tool if necessary. I don't see this as a step back. It will be clearly visible what tools the LLM uses.

<img width="978" height="484" alt="image" src="https://github.com/user-attachments/assets/7fd2b146-0777-43bb-88b4-5cc2d2d6690b" />


By continuing the chat, you can receive answers in the same context and receive calls to tools

![calc](https://github.com/user-attachments/assets/8c8f2d0a-ddec-4b6a-965e-b6f78548e6ba)

paymicro committed:
Enhance Turbo chat tool integration and UI

Refactored Turbo chat controls to use toggle buttons for API and SQL tool context, improved message handling and rendering, and added support for showing tool calls. Updated Conversation logic for tool management, message checkpointing, and function call handling. Improved SQL Server agent feedback and streamlined UI elements for better accessibility and usability.

paymicro commented on VisualChatGPTStudioShared/ToolWindows/Turbo/TerminalWindowTurboControl.xaml:
```diff
@@ -270,163 +307,77 @@ Padding="8" CornerRadius="4">
                     ToolTip="Attach an image"
                     AutomationProperties.HelpText="Attach an image"
                     TabIndex="4"
-                    IsTabStop="True"
```
`IsTabStop` moved to button style. All buttons already have this property. Removed duplication

paymicro committed:
Add DataTable to Markdown conversion and refactor

Introduced Extensions.cs with a method to convert DataTable to Markdown format. Refactored TerminalWindowTurboControl.xaml.cs to use this method for displaying SQL query results in chat. Moved RelayCommand.cs to Turbo folder and updated project file references accordingly.

paymicro committed:
table css style

paymicro committed:
Refactor SQL Server Agent data handling to use dictionaries

Replaces DataTable/DataView usage with List for SQL query results in SqlServerAgent and related UI code. Updates markdown rendering extension and UI logic to work with the new data structure. Simplifies data flow and improves compatibility with non-DataTable environments. Also aligns table cell text to the left in TurboChat.css.

paymicro committed:
Update message handling for API responses

Changed message segment author from ChatGPT to Api for DB results, updated view model to skip loading Api messages to conversation. Removed unused helper methods.

jeffdapaz commented on VisualChatGPTStudioShared/ToolWindows/Turbo/TerminalWindowTurboControl.xaml.cs:
```diff
@@ -647,53 +755,60 @@ private async Task<bool> HandleFunctionsCallsAsync(List<FunctionResult> function
-                apiChat.AppendToolMessage(function.Id, functionResult);
+                _viewModel.ToolCallAttempt++;
+
+                _viewModel.apiChat.AppendToolMessage(function, functionResult);
```
>Please give an example

I believe there are two types of solution:

1 – Keep the rendering via markdown, but limit the table display with a defined height and width. The height would have a predefined value, while the width would be adjusted according to the width of the chat window, similar to what already happens with code blocks. Navigation would be done via scrollbars.

2 – Use a free library via CDN to render the table. Examples:

**[DataTables](https://diegomariano.com/datatables/):**

```html
<link rel="stylesheet" href="https://cdn.datatables.net/1.13.6/css/jquery.dataTables.min.css">
<script src="https://code.jquery.com/jquery-3.7.0.min.js"></script>
<script src="https://cdn.datatables.net/1.13.6/js/jquery.dataTables.min.js"></script>

<table id="myTable"> ... </table>
<script>
  $(document).ready(function(){ $('#myTable').DataTable({ pageLength: 25 }); });
</script>
```

<img width="1200" height="641" alt="image" src="https://github.com/user-attachments/assets/fcef44fe-3bea-4aef-8cfb-43f262f92803" />



**[Grid.js](https://gridjs.io/)**

```html
<link href="https://unpkg.com/gridjs/dist/theme/mermaid.min.css" rel="stylesheet"/>
<script src="https://unpkg.com/gridjs/dist/gridjs.umd.js"></script>

<div id="wrapper"></div>
<script>
  new gridjs.Grid({
    columns: ["Nome","Idade","País"],
    data: [["João",30,"PT"],["Ana",28,"BR"]],
    pagination: { limit: 10 },
    sort: true,
    search: true
  }).render(document.getElementById("wrapper"));
</script>
```

<img width="483" height="385" alt="image" src="https://github.com/user-attachments/assets/b3ded459-d616-432f-9124-593d98f218a4" />

Personally, I prefer using a library because it is more feature-rich, offering options like pagination and ordering, for example.

paymicro committed:
gridjs for tables

jeffdapaz commented:
@paymicro,

Will you submit a new commit, or have you already finished all pendings?

When you're done, I'll grab the latest version to run tests again.

paymicro committed:
minor refactoring

paymicro committed:
major refactoring - move logic to ViewModel

paymicro committed:
image + tools in one request

paymicro committed:
fix message duplication

paymicro committed:
fix azure

jeffdapaz commented:
@paymicro, with your latest fix I was able to test. I ran a few tests and found the issues below; however, I haven’t tested the functions yet (API, SQL, and Computer User)—I’ll get to that later.

1 - _“Do not automatically scroll down while receiving the streaming, as it forces us to scroll up to read the beginning of the message.”_

It still scrolls to the end of the message.

2 - _“Reintroduce the ability to edit the chat name on the history screen.”_

It’s still not possible to confirm the chat name change.

3 - _“When opening an existing chat, ideally position the scroll at the beginning of the last response, so it's not necessary to scroll up to read the last message. This is already the case in the current version.”_

It still scrolls to the end of the message.

4 - When clicking the “Copy as image” button on a mermaid diagram, provide feedback similar to the other commands that already do this.

5 - _“Highlight special items: ‘/’ for commands, ‘@’ for items, and API and SQL tags within the chat, just like in the current version.”_

Still not highlighting.

6 - Referencing parts of the project using “@” stopped working. With “@” in the request field, the window to reference parts of the project is not being displayed.

7 - _“In the chat (selected list item), allow choosing the ‘Rename’ and ‘Delete’ commands with the TAB key, and confirm and execute the button with Enter.”_

It’s now possible to choose between “Rename” and “Delete” using the TAB key, but when trying to confirm one of the selected commands, pressing ENTER opens the chat instead of executing the button’s function.

8 - _“The shortcut CTRL+ALT+Enter for the ‘Send Code’ command is not working.”_

Still not working.

9 - _“In the current version of the chat, when the ‘Send Code’ button is pressed, it checks if the AI’s response contains a code block, and if it does, it applies this block to the active editor, and what is not a code block is printed in the chat window. If the ‘Send Code’ command is pressed while holding the SHIFT key, it will not write the code block from the response to the editor, but rather in the chat. However, from what I tested, now it always writes the code block in the chat, regardless of whether I pressed the SHIFT key or not.”_

Still not working.

10 - When trying to open the Turbo Chat window in VS2019, nothing happens—the window doesn’t open and no error is triggered. However, I still can’t debug the VS2019 version, and I don’t know why. What I had to do to be able to test was install the VSIX generated in the release folder.

paymicro commented:
VS 2019 Exception. It looks like the library for `System.Text.Json` is missing. I'll fix it soon
```
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.Windows.Markup.XamlParseException: 'The invocation of the constructor on type 'VisualChatGPTStudioShared.ToolWindows.Turbo.TerminalTurboViewModel' that matches the specified binding constraints threw an exception.' Line number '20' and line position '10'. ---> System.IO.FileNotFoundException: Could not load file or assembly 'System.Memory, Version=4.0.1.2
```

paymicro committed:
fix VS 2019

paymicro committed:
fix chat renaming

paymicro committed:
fix key handling in history buttons

Adds support for executing button commands with Enter or Space keys and clarifies key handling logic for editing and navigation in the history list.

paymicro committed:
SendCode in stream + highlight Special Tags

paymicro commented:
> When running the extension in debug mode, I receive this error in VS2026 right at the beginning:


This error also occurs in 2022 when debugging. 
Just reset experimental Instance for 2026 and remove folder `18.0_` + blabla123 + `Exp` in `%AppData%\Microsoft\VisualStudio\`

paymicro committed:
use WebView2 for all versions

paymicro committed:
dont load tool role from history

jeffdapaz commented:
>Perhaps this is after loading the chat with calls to tools through history and continuing the dialogue. 

@paymicro,

I think it's worth taking a look at the current code, as currently always loading from history, and the last time I reloaded a chat with calls to Functions, it was working without problems:

<img width="874" height="576" alt="image" src="https://github.com/user-attachments/assets/7b25c43d-d8d4-4a26-87c5-422a2741603e" />

jeffdapaz commented:
> <img alt="image" width="564" height="115" src="https://private-user-images.githubusercontent.com/27482193/513991247-24782678-aa2d-4fc9-a74c-c03d51dc0bcf.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjMwNTcwMDAsIm5iZiI6MTc2MzA1NjcwMCwicGF0aCI6Ii8yNzQ4MjE5My81MTM5OTEyNDctMjQ3ODI2NzgtYWEyZC00ZmM5LWE3NGMtYzAzZDUxZGMwYmNmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTEzVDE3NTgyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThiZDAxZmNkMjZkNDNlMWRjNjYzZGVkMjgxZDIwNTIwNjVlMzQ5MzU1YTc3YjE4OGJlZTQ0ODllOTc2YWJkM2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mjhMNQvslKssOCo738XBtbhQWrzBisEFR_cR6aCJxc0">
> At the current code I don't see Assistants messages in `apiChat`. ChatGpt author sends as "user"

I don't understand why you're mentioning Assistants messages. Especially since the error refers to the "tool" role, which are the function definitions.

<img width="854" height="143" alt="image" src="https://github.com/user-attachments/assets/46a06094-c76e-4458-a518-3000c8df3b07" />

<img width="627" height="154" alt="image" src="https://github.com/user-attachments/assets/566e6bb5-0f9f-429b-9b50-f68d6c8d073e" />

And in your last commit, it's not reloading the functions when reloading a chat, whereas this currently works:

<img width="856" height="136" alt="image" src="https://github.com/user-attachments/assets/a6d7b474-abc0-4da3-9f82-37fd0870ee72" />

paymicro commented:
> When you reload the chat, messages with the Assistance role are not reloaded; in fact

Yes. Let's finish discussing this problem. It seems we understood each other and there is no problem now

paymicro committed:
fix tool calls

paymicro committed:
support Disable certificate validation

paymicro committed:
Add system message tool support and BuiltInAgent

Introduces experimental support for system message tools, including a new BuiltInAgent for handling built-in tool calls. Updates Conversation and TerminalTurboViewModel to support system message tool workflows, adds related options, and refactors tool handling logic. Improves UI feedback in TurboChat.js and TurboControl.xaml, and fixes minor issues in code completion and agent APIs.

paymicro committed:
Refactor tool management and improve UI feedback

Introduced a Tool class and centralized tool definitions in BuiltInAgent for better maintainability. Updated GetToolUseInstructions to dynamically generate instructions from tool definitions. Improved TurboChat.js to handle missing SVG elements and provide error feedback in the UI. Fixed minor logic issues in JavaScript popup messages and button swapping.

paymicro committed:
Refactor tool and agent architecture

Introduces new Tool, ToolManager, ToolResult, ApprovalKind, and RiskLevel abstractions to support a unified and extensible tool execution model. Refactors BuiltInAgent, ApiAgent, and SqlServerAgent to use strongly-typed tool definitions and async execution, replacing previous function-based logic. Updates Conversation and related classes to support new tool invocation and parsing. Removes unused properties and streamlines code for maintainability and future feature expansion.

paymicro committed:
Track response model and fix nullable index handling

Adds a ResponseModel property to Conversation to store the model used in API responses. Updates FunctionResult.Index to be nullable and ensures correct handling of null index values in tool calls.

paymicro committed:
restore support VS2019

paymicro committed:
fix read files

paymicro committed:
fix tools

paymicro committed:
fix send request

paymicro committed:
change format for tools in system message

jeffdapaz commented:
Hi @paymicro,

I grabbed your latest commit to run some tests, and while other commands and windows I thinking are now working fine, with Turbo Chat I encountered an error on my very first message.

<img width="306" height="138" alt="image" src="https://github.com/user-attachments/assets/e4a18538-3c38-4c1b-ae80-7eab67c6e164" />

paymicro committed:
Refactor tool APIs and improve function handling

Reworked built-in agent tools: replaced file_glob_search with file_search using regex, standardized grep_search to use regex, and improved parameter definitions. Refactored solution file discovery and moved related helpers. Updated function call handling in TerminalTurboViewModel for better async flow and UI updates. Removed unused Markdig dependencies from project files.

jeffdapaz commented:
Hi @paymicro,

With this latest version, I was able to run the tests. I’d say it’s already quite stable and close to being ready for a release. I’d also like to commend you for the excellent improvements. Here’s what I’ve found so far:

1 - When clicking the “Cancel” button, the action isn’t immediately interrupted (the progress bar still remains active), and at the end, an “object reference not set...” error is displayed.

2 - The vertical scroll bar still moves to the end of the chat while streaming is in progress, whereas ideally, it should stay at the beginning of the response so the user doesn’t have to scroll up to view the start of the message.

3 - Why did you abandon the use of WebView2CompositionControl? This way, the chat window is overlapping Visual Studio’s own windows.

4 - The shortcut CTRL+ALT+ENTER for the “Send Code” button isn’t working.

5 - The progress bar is only displayed for the “Send” command. For all other commands, when waiting for the API response, the progress bar isn’t shown.

6 - When I tried using the “Send Code” command, it didn’t work because the class code wasn’t sent:

```json
    {
        "model": "gpt-3.5-turbo",
        "messages": [
            {
                "role": "system",
                "content": "You are a programmer assistant called Visual chatGPT Studio, and your role is help developers and resolve programmer problems."
            },
            {
                "role": "user",
                "content": "<task>Add a new method to this class</task> <code> </code>"
            }
        ],
        "n": 1,
    ...
```

7 - Even after adding an API to the chat context, the API combo box still remained visible.

By the way, one question: How does the “BuiltInAgent” work? I see that it can manipulate the project in various ways, similar to GitHub Copilot, but how and when will it be invoked?

P.S.: I just haven’t had time to test the SQL Server Agent yet. I’ll do that as soon as possible.

Commit:
Improve GetConnections to extract real connection string when connection is protected with password.

paymicro committed:
Refactor tool parameter handling and add new features

Replaces 'Parameters' with 'Properties' for tool definitions and updates related logic to support more flexible parameter schemas. Adds new built-in tools for building solutions and retrieving error lists, refactors file path handling to support absolute paths, and improves terminal command execution with explicit 'exe' selection. Updates UI layouts and improves JSON serialization options for better compatibility.

paymicro committed:
minor fixing for tools
